materialize-mongodb/transactor.go

```diff
@@ -6,143 +6,279 @@ import (
+		if lastBinding != -1 && (lastBinding != it.Binding || len(batch) == storeBatchSize || batchSize >= batchByteLimit) {
+			if err := sendBatch(); err != nil {
+				return nil, err
+			}
+		}
+
+		key := fmt.Sprintf("%x", it.PackedKey) // Hex-encode
 
 		var doc bson.M
 		if err := json.Unmarshal(it.RawJSON, &doc); err != nil {
 			return nil, fmt.Errorf("bson unmarshalling json doc: %w", err)
 		}
-		// In case of delta updates, we don't want to set the _id. We want
-		// MongoDB to generate a new _id for each record we insert
-		if !binding.res.DeltaUpdates {
+		// In case of delta updates, we don't want to set the _id. We want MongoDB to generate a new
+		// _id for each record we insert
+		if !t.bindings[it.Binding].deltaUpdates {
 			doc[idField] = key
 		}
 
+		var m mongo.WriteModel
 		if it.Exists {
-			var opts = options.Replace()
-			_, err := binding.collection.ReplaceOne(tx, bson.D{{idField, bson.D{{"$eq", key}}}}, doc, opts)
-			if err != nil {
-				return nil, fmt.Errorf("upserting document into collection %s: %w", binding.collection.Name(), err)
+			m = &mongo.ReplaceOneModel{
+				Filter:      bson.D{{Key: idField, Value: bson.D{{Key: "$eq", Value: key}}}},
+				Replacement: doc,
 			}
 		} else {
-			_, err := binding.collection.InsertOne(tx, doc, options.InsertOne())
-			if err != nil {
-				return nil, fmt.Errorf("inserting document into collection %s: %w", binding.collection.Name(), err)
-			}
+			m = &mongo.InsertOneModel{Document: doc}
 		}
+
+		batch = append(batch, m)
+		batchSize += len(it.RawJSON)
+		lastBinding = it.Binding
 	}
 
-	var filter = bson.D{{"materialization", bson.D{{"$eq", t.fence.Materialization}}}, {"fence", bson.D{{"$eq", t.fence.Fence}}}}
-	var fence fenceRecord
-	if err = t.fenceCollection.FindOne(tx, filter, options.FindOne()).Decode(&fence); err != nil {
-		return nil, fmt.Errorf("finding fence: %w", err)
+	// Drain the last batch.
+	if err := sendBatch(); err != nil {
+		return nil, err
 	}
 
-	return func(ctx context.Context, runtimeCheckpoint *protocol.Checkpoint, runtimeAckCh <-chan struct{}) (*pf.ConnectorState, pf.OpFuture) {
-		checkpointBytes, err := runtimeCheckpoint.Marshal()
-		if err != nil {
-			return nil, pf.FinishedOperation(fmt.Errorf("marshalling checkpoint: %w", err))
-		}
+	close(sendBatches)
+	return nil, group.Wait()
+}
+
+func (t *transactor) Destroy() {}
+
+type loadBatch struct {
+	binding int
+	keys    []string
+}
 
-		return nil, pf.RunAsyncOperation(func() error {
-			var bump = bson.D{{"$set", bson.D{{"checkpoint", checkpointBytes}}}}
-			var updateOpts = options.Update()
-			if _, err = t.fenceCollection.UpdateOne(ctx, filter, bump, updateOpts); err != nil {
-				return fmt.Errorf("updating checkpoint: %w", err)
+func (t *transactor) loadWorker(
+	ctx context.Context,
+	loaded func(i int, doc json.RawMessage) error,
+	batches <-chan loadBatch,
+) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case batch := <-batches:
+			if batch.keys == nil {
```

Comment:
Trivial nit: If you're just checking whether the channel is closed I think `case batch, ok := <-batches: if !ok ...` is more idiomatic. But this definitely works so feel free to ignore.

Reply:
Oh nice, that is better. It felt a little weird checking for the zero-value.